rfcs/text/0003_handler_interface.md

```diff
@@ -0,0 +1,173 @@
+different solutions are developed per handler for managing complexity and
+enabling extensibility.
+
+The official handler interface seeks to address five key problems:
+
+1. Different services and plugins should be able to expose functionality that
+   is configured for the particular context where the handler is invoked, such
+   as a savedObject client in an alert handler already being configured to use
+   the appropriate API token.
+
+2. The service owner of a handler should not need to know about the services
+   or plugins that extend its handler context, such as the security plugin
+   providing a currentUser function to an HTTP router handler.
+
+3. Many capabilities in a handler should be "fixed" for the life of that
+   handler's context rather than changing configuration under the hood in
+   mid-execution. For example, while Elasticsearch clients can technically
+   be replaced throughout the course of the Kibana process, an HTTP route
+   handler should be able to depend on their being a consistent client for its
+   own shorter lifespan.
+
+4. Plugins should not need to pass down high level service contracts throughout
+   their business logic just so they can access them within the context of a
+   handler.
+
+5. Many capabilities provided by services should not be arbitrarily used in
+   unconstrained execution such as in the plugin lifecycle hooks. For example,
+   it's appropriate for an Elasticsearch client to throw an error if it's used
+   inside an API route and Elasticsearch isn't available, however it's not
+   appropriate for a plugin to throw an error in their start function if
+   Elasticsearch is not available. If the ES client was only made available
+   within the handler context and not to the plugin's start contract at large,
+   then this isn't an issue we'll encounter in the first place.
+
+# Detailed design
+
+There are two parts to this proposal. The first is the handler interface
+itself, and the second is the interface that a service owner implements to make
+their handlers extensible.
+
+## Handler interface
+
+```ts
+interface HandlerContext {
+  core?: Record<string, unknown>;
+  [serviceName: string]: Record<string, unknown>;
+}
+
+type Handler = async (HandlerContext, ...args: unknown[]) => unknown;
+```
+
+## Handler creator interface
+
+```ts
+type CapabilityCreator = async (...args: unknown[]) => Record<string, unknown>;
+
+interface HandlerCreator {
+  registerCapability(serviceName: string, CapabilityCreator);
+}
+```
+
+Note: The `HandlerCreator` is defined as a literal interface in this document,
+but in practice this interface is just a guide for the pattern of registering
+capabilities. Certain services may have multiple different types of handlers,
+so they may choose not to use the generic name `registerCapability` in favor of
+something more explicit.
+
+## End to end example
+
+```js
+http.router.registerRequestCapability('elasticsearch', async request => {
```

Comment:
if functionality is exposed via `registerRequestCapability`, will it be available via Setup/start contracts?